lib/autoupdate/agent/validate.go

```diff
@@ -0,0 +1,130 @@
+	// fileHeaderSniffBytes is the max size to read to determine a file's MIME type
+	fileHeaderSniffBytes = 512 // MIME standard size
+	// execModeMask is the minimum required set of bits to consider a file executable.
+	execModeMask = 0111
+)
+
+// Validator validates filesystem paths.
+type Validator struct {
+	Log *slog.Logger
+}
+
+// IsBinary returns true for working binaries that are executable by all users.
+// If the file is irregular, non-executable, or a shell script, IsBinary returns false and logs a warning.
+// IsBinary errors if lstat fails, a regular file is unreadable, or an executable file does not execute.
+func (v *Validator) IsBinary(ctx context.Context, path string) (bool, error) {
+	// The behavior of this method is intended to protect against executable files
+	// being adding to the Teleport tgz that should not be made available on PATH,
+	// and additionally, should not cause installation to fail.
+	// While known copies of these files (e.g., "install") are excluded during extraction,
+	// it is safer to assume others could be present in past or future tgzs.
+
+	if exec, err := v.IsExecutable(ctx, path); err != nil || !exec {
+		return exec, trace.Wrap(err)
+	}
+	name := filepath.Base(path)
+	d, err := readFileLimit(path, fileHeaderSniffBytes)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	// Refuse to test or link shell scripts
+	if isTextScript(d) {
+		v.Log.WarnContext(ctx, "Found unexpected shell script", "name", name)
+		return false, nil
+	}
+	v.Log.InfoContext(ctx, "Validating binary", "name", name)
+	r := localExec{
+		Log:      v.Log,
+		ErrLevel: slog.LevelDebug,
+		OutLevel: slog.LevelInfo, // always show version
+	}
+	code, err := r.Run(ctx, path, "version")
+	if code < 0 {
+		return false, trace.Wrap(err, "error validating binary %s", name)
+	}
+	if code > 0 {
+		v.Log.InfoContext(ctx, "Binary does not support version command", "name", name)
+	}
+	return true, nil
+}
+
+// IsExecutable returns true for regular, executable files.
+func (v *Validator) IsExecutable(ctx context.Context, path string) (bool, error) {
+	name := filepath.Base(path)
+	fi, err := os.Lstat(path)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	if !fi.Mode().IsRegular() {
+		v.Log.WarnContext(ctx, "Found unexpected irregular file", "name", name)
+		return false, nil
+	}
+	if fi.Mode()&execModeMask != execModeMask {
+		v.Log.WarnContext(ctx, "Found unexpected non-executable file", "name", name)
+		return false, nil
+	}
+	return true, nil
+}
+
+func isTextScript(data []byte) bool {
+	data = bytes.TrimLeftFunc(data, unicode.IsSpace)
+	if !bytes.HasPrefix(data, []byte("#!")) {
+		return false
+	}
+	// Assume presence of MIME binary data bytes means binary:
+	//   https://mimesniff.spec.whatwg.org/#terminology
+	for _, b := range data {
+		switch {
+		case b <= 0x08, b == 0x0B,
+			0x0E <= b && b <= 0x1A,
+			0x1C <= b && b <= 0x1F:
+			return false
+		}
+	}
+	return true
+}
+
+// readFileLimit the first n bytes of a file.
```

Comment:
Isn't this the same as `readFileAtMost`?

Reply:
`readFileAtMost` errors if the file is too large (behavior of `utils.ReadAtMost`), while `readFileLimit` does not.